sklearn/cluster/_k_means_elkan.pyx

```diff
@@ -0,0 +1,176 @@
+
+    Parameters
+    ----------
+    X_ : nd-array, shape (n_samples, n_features)
+
+    n_clusters : int
+        Number of clusters to find.
+
+    init : nd-array, shape (n_clusters, n_features)
+        Initial position of centers.
+
+    tol : float, default=1e-4
+        The relative increment in cluster means before declaring convergence.
+
+    max_iter : int, default=30
+        Maximum number of passes over the dataset.
+
+    verbose : bool, default=False
+        Whether to be verbose.
+
+    """
+    #initialize
+    centers_ = init
+    cdef double* centers_p = getfloatpointer(centers_)
+    cdef double* X_p = getfloatpointer(X_)
+    cdef double* x_p
+    cdef int n_samples = X_.shape[0]
+    cdef int n_features = X_.shape[1]
+    cdef int point_index, center_index, label
+    cdef float upper_bound, distance
+    cdef double[:, :] center_distances = euclidean_distances(centers_) / 2.
+    cdef double[:, :] lower_bounds = np.zeros((n_samples, n_clusters))
+    cdef double[:] distance_next_center
+    labels_ = np.empty(n_samples, dtype=np.int32)
+    cdef int[:] labels = labels_
+    upper_bounds_ = np.empty(n_samples, dtype=np.float)
+    cdef double[:] upper_bounds = upper_bounds_
+    update_labels_distances_inplace(X_p, centers_p, center_distances, labels,
+                                    lower_bounds, upper_bounds, n_samples,
+                                    n_features, n_clusters)
+    cdef np.uint8_t[:] bounds_tight = np.ones(n_samples, dtype=np.uint8)
+    cdef np.uint8_t[:] points_to_update = np.zeros(n_samples, dtype=np.uint8)
+
+    if max_iter <= 0:
+        raise ValueError('Number of iterations should be a positive number'
+        ', got %d instead' % max_iter)
+
+    for iteration in range(max_iter):
+        if verbose:
+            print("start iteration")
+        # we could find the closest center in O(n) but
+        # this does not seem to be the bottleneck
+        distance_next_center = np.sort(center_distances, axis=0)[1]
+        if verbose:
+            print("done sorting")
+        for point_index in range(n_samples):
+            upper_bound = upper_bounds[point_index]
+            label = labels[point_index]
+            if distance_next_center[label] >= upper_bound:
+                continue
+            x_p = X_p + point_index * n_features
+            # check other update conditions
+            # TODO: get pointer to lower_bounds[point_index, center_index]
+            for center_index in range(n_clusters):
+                if (center_index != label
+                        and (upper_bound > lower_bounds[point_index, center_index])
+                        and (upper_bound > center_distances[center_index, label])):
+                    # update distance to center
+                    if not bounds_tight[point_index]:
+                        upper_bound = d(x_p, centers_p + label * n_features, n_features)
+                        lower_bounds[point_index, label] = upper_bound
```

Comment:
It seems that this should be `upper_bounds[point_index, label]` no? because the distance between the current center and the sample is recalculated?
See 3a in Algorithm section:
`u(x)` is set to `d(x, c(x)`